### PR TITLE
Rest endpoint sysl

### PR DESCRIPTION
**Pull request by nofun97**
Will try to work on `simple_endpoint` too but so far, it should parse `rest_endpoint` correctly.

### CODE REVIEW

File: examples/sysl.wbnf [hunk 1]
```diff
@@ -45,18 +56,76 @@ view_params -> (NAME "<:" (collection_type|type_spec)):",";
...
-PATH    -> /{[a-zA-Z0-9.:]*}:"/";
+TEXT_LINE -> /{ \s* ( [^\s(] [^(\r\n]* [^\s(] ) \s* };
+DIGITS -> /{[1-9][0-9]*};
+TEXT -> /{[^(\r?\n)]+};
+PATH    -> /{[a-zA-Z0-9.]*}:"/";
 APP_NAME -> NAME:".";
+PREDICATE -> /{[^\r\n:]+};
 SHORTCUT -> "...";
-
+HTTP_VERBS -> ("GET" | "POST" | "DELETE" | "PUT" | "PATCH");
```
Comment (anzdaddy) [inline examples/sysl.wbnf]: Parentheses not required

File: examples/sysl.wbnf [hunk 2]
```diff
@@ -45,18 +56,76 @@ view_params -> (NAME "<:" (collection_type|type_spec)):",";
...
+NativeDataTypes -> ("int32" | "int64" | "int" |
```
Comment (anzdaddy) [inline examples/sysl.wbnf]: Parentheses not required

File: examples/sysl.wbnf [hunk 3]
```diff
@@ -45,18 +56,76 @@ view_params -> (NAME "<:" (collection_type|type_spec)):",";
 
 expr_block -> "";
 
+// --------------  statements  --------------- //
+statements -> (if_stmt
```
Comment (anzdaddy) [inline examples/sysl.wbnf]: `stmt -> ...`

File: examples/sysl.wbnf [hunk 4]
```diff
@@ -45,18 +56,76 @@ view_params -> (NAME "<:" (collection_type|type_spec)):",";
...
+
+if_stmt -> "if" PREDICATE ":" INDENT statements*:INDENT_SEP;
+
+else_stmt -> "else" PREDICATE? ":" INDENT statements*:INDENT_SEP;
+
+for_stmt -> for_type=("alt" | "until" | "for each" | "for" | "loop" | "while") PREDICATE ":" INDENT statements*:INDENT_SEP;
+
+one_of_stmt -> "one" "of" ":" INDENT one_of_cases+:INDENT_SEP;
+
+one_of_cases -> one_of_case_label=(NAME | TEXT_LINE | QSTRING)? ":" INDENT statements+:INDENT_SEP;
```
Comment (anzdaddy) [inline examples/sysl.wbnf]: Is the `+` required? Ditto elsewhere.

**Review by anzdaddy — CHANGES_REQUESTED** [review]

File: examples/sysl.wbnf (hunk 4 above)
Comment (nofun97) [reply examples/sysl.wbnf]: just realized it's wrong to use that, will change all of them

**Comment by nofun97** [comment]
working endpoints

File: examples/sysl.wbnf [hunk 5]
```diff
@@ -10,60 +10,131 @@ import  -> "import"  prefix="/"{0,2} import=PATH ( "as"  APP_NAME)? ("~" mode=NA
...
+
+for_stmt -> for_type=("alt" | "until" | "for each" | "for" | "loop" | "while") PREDICATE ":" \n+ level=INDENT stmt:INDENT_SEP;
+
+one_of_stmt -> "one" "of" ":" \n+ level=INDENT one_of_cases:INDENT_SEP;
+
+one_of_cases -> one_of_case_label=(QSTRING ":"| NAME ":" | TEXT_LINE ":" | ":")  \n+ level=INDENT stmt:INDENT_SEP;
+
+call_stmt -> target=("." | NAME:"::") "<-" target_endpoint=(TEXT_LINE) call_args?;
+
+call_args -> "(" ((QSTRING | NAME)+ | (NAME "<:" (NAME | NativeDataTypes))):"," ")";
```
Comment (ghost) [inline examples/sysl.wbnf]: `(NAME | NativeDataTypes)` should be `type_spec` check elsewhere also

File: examples/sysl.wbnf [hunk 6]
```diff
@@ -10,60 +10,131 @@ import  -> "import"  prefix="/"{0,2} import=PATH ( "as"  APP_NAME)? ("~" mode=NA
...
+
 // -------------- Common Rules --------------- //
 attribs     ->  "[" @:"," "]"
              > pattern=("~" NAME:"+")
              | attr=(NAME "=" QSTRING)
         ;
 
+array_size -> "(" DIGITS ".." DIGITS? ")";
+
+reference -> package_name=(NAME:"::") ("." NAME)+;
```
Comment (ghost) [inline examples/sysl.wbnf]: APPLICATION_NAME should be used here too

File: examples/sysl.wbnf [hunk 7]
```diff
@@ -10,60 +10,131 @@ import  -> "import"  prefix="/"{0,2} import=PATH ( "as"  APP_NAME)? ("~" mode=NA
...
-    INDENT
-        (table | "!table"APP_NAME) :INDENT_SEP EOL;
+    \n+ level=INDENT
+        (table | "!table"APP_NAME):INDENT_SEP EOL;
 
 
-type_spec -> (NativeDataType=("int32" | "int64" | "int" |
-                             "float" | "string" | "date" |
-                             "bool" | "decimal" | "datetime") | UserType=APP_NAME)
+type_spec -> (NativeDataTypes | UserType=APP_NAME)
```
Comment (ghost) [inline examples/sysl.wbnf]: Reference should be an option also

**Review by ghost — COMMENTED** [review]
couple more nitpicks, looking good (probably more room to simplify also)

File: examples/sysl.wbnf (hunk 5 above)
Comment (nofun97) [reply examples/sysl.wbnf]: `type_spec` allows `APP_NAME`, is that okay for `call_args`?

Comment (nofun97) [reply examples/sysl.wbnf]: should I change it to 
```
type_spec -> (NativeDataTypes | UserType=APPLICATION_NAME | NAME)
             SizeSpec=("(" \d+ (array=".."|sized=".")? \d* ")")?;
```

File: examples/sysl/endpoints.wbnf [hunk 8]
```diff
@@ -0,0 +1,18 @@
+// -------------- Endpoint     --------------- //
```
Comment (ghost) [inline examples/sysl/endpoints.wbnf]: remove these headers if we are splitting up the grammar into multiple files

File: examples/wbnf.wbnf [hunk 9]
```diff
@@ -40,10 +40,8 @@ RE      -> /{
                )(?: (?:[+*?]|\{\d+,?\d?\}) \?? )?
              )+
            };
-REF     -> "%" IDENT ("=" default=STR)?;
-// Special
 pragma  -> (
                 import=(".import" path=((".."|"."|[a-zA-Z0-9.:]+):,"/"))
-           );
+REF     -> "%" IDENT ("=" default=STR)?;
```
Comment (ghost) [inline examples/wbnf.wbnf]: revert this change, looks like a bad merge

**Review by ghost — COMMENTED** [review]
minor nits, good enough for now i think

File: examples/wbnf.wbnf (hunk 9 above)
Comment (nofun97) [reply examples/wbnf.wbnf]: reverted

File: examples/sysl/endpoints.wbnf (hunk 8 above)
Comment (nofun97) [reply examples/sysl/endpoints.wbnf]: removed
